lightning/src/chain/keysinterface.rs

```diff
@@ -1236,14 +1231,16 @@ impl KeysManager {
 
 impl EntropySource for KeysManager {
 	fn get_secure_random_bytes(&self) -> [u8; 32] {
-		let mut sha = self.rand_bytes_unique_start.clone();
+		let index = self.rand_bytes_index.fetch_add(1, Ordering::AcqRel);
 
-		let child_ix = self.rand_bytes_child_index.fetch_add(1, Ordering::AcqRel);
-		let child_privkey = self.rand_bytes_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32).expect("key space exhausted")).expect("Your RNG is busted");
-		sha.input(&child_privkey.private_key[..]);
+		let secs_plus_index = self.starting_time_secs + index as u64;
+		let mut nonce = Vec::new();
+		nonce.append(&mut secs_plus_index.to_be_bytes().to_vec());
+		nonce.append(&mut self.starting_time_nanos.to_be_bytes().to_vec());
 
-		sha.input(b"Unique Secure Random Bytes Salt");
-		Sha256::from_engine(sha).into_inner()
+		let mut random_bytes = [0; 32];
+		ChaCha20::new(&self.seed, &nonce).process_in_place(&mut random_bytes);
```

Comment:
Rather than creating a new `ChaCha20` instance on each call, can't we keep the state in a corresponding field once and here only do the `process_in_place` part? Additionally, this would allow us to get rid of the iterating `index`.

Reply:
Applied in [d3f14e7](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d3f14e7e2951892a70f72a4f2234416e1a10cb1f)

Reply:
Thanks for the fix! While it seems to be the most straightforward,  I unfortunately suspect that taking a `Mutex` lock is likely too slow and we'd probably want to explore alternatives, e.g. the one sketched above.

As this PR aims to improve perfomance, could you add a benchmark test that would allow us to compare the approaches head-to-head (old vs. Mutex vs. iterating nonce)?

Reply:
True. I've also discovered that the `Mutex` isn't even a real option due to the `no-std` feature. I will work on the benchmark test to compare the approaches but replace the Mutex one with a simple spin-lock.

Reply:
> True. I've also discovered that the `Mutex` isn't even a real option due to the `no-std` feature. I will work on the benchmark test to compare the approaches but replace the Mutex one with a simple spin-lock.

Ah, but regarding that you could just use `crate::sync::Mutex`, which chooses the right one based on `std`/`no-std`. Still, probably not the way to go.

Reply:
Ok, I've found the time to come back to this. I've only noticed now that I didn't read quite well your first message. That's why I just went ahead with the `Mutex` based approach...

I've added the benchmark test in [e3622f1](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/e3622f136a886dfc84242d7a9872f2f6d1aded8b), followed by commit [fac7b4e](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/fac7b4e02777819868f68d539506ef70ed6ed36a) with the `Mutex` approach, finally followed by [d943b28](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d943b2896e25b75c4b13ffcba8982da9e86ccc1e) with the iterating nonce approach. On my machine, the `Mutex` ends up being faster than the iterating nonce one by about 30-40%, but I'll wait for the CI to run to confirm my observation :)

Reply:
I should have thought of that! Now the results make more sense :) The `Mutex` approach is 3x slower when 5 threads compete for the lock. 

Values on my machine for 100 calls:
- Single-thread `Mutex` -> ~ 6 us
- Single-thread iterating nonce -> ~10 us
- 5 competing threads `Mutex` -> ~30 us
- 5 competing threads iterating nonce -> ~90 us
- 5 competing threads original approach -> ~1900 us

It's a trade-off between getting a higher throughput implementation in high vs. low contention scenarios. I think it makes sense to go with the iterating nonce approach. Despite not reaching the highest throughput, it achieves significantly higher throughput when there's more competition.

I've pushed [3266726](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/326672612291e59b0612dcc71ff682d6cb3383cd) with the updated benchmark and [17860ac](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/17860ac0f85b82ad4df761a6ef0cd996903c1229) with the iterating nonce approach.

Thank you for the feedback and help.

With the new implementation, other tests are failing, maybe for the same reason as the one addressed in #1984. I will have a look later.

Reply:
I'm sorry, somehow I switched the 2 values by mistake 🤦 I edited the comment now to get it fixed. I will push the 2 versions soon, I have them locally.

Reply:
> I will push the 2 versions soon, I have them locally.

Done!